pkg/controller/inferenceservice/resources/istio/virtualservice_test.go

```diff
@@ -0,0 +1,466 @@
+		expectedStatus  *v1alpha2.VirtualServiceStatus
+		expectedService *istiov1alpha3.VirtualService
+	}{{
+		name:            "nil status should not be ready",
+		defaultStatus:   nil,
+		canaryStatus:    nil,
+		expectedStatus:  createFailedStatus(PredictorStatusUnknown, PredictorMissingMessage),
+		expectedService: nil,
+	}, {
+		name:            "empty status should not be ready",
+		defaultStatus:   &v1alpha2.ComponentStatusMap{},
+		canaryStatus:    nil,
+		expectedStatus:  createFailedStatus(PredictorStatusUnknown, PredictorMissingMessage),
+		expectedService: nil,
+	}, {
+		name: "predictor missing host name",
+		defaultStatus: &v1alpha2.ComponentStatusMap{
+			constants.Predictor: &v1alpha2.StatusConfigurationSpec{},
+		},
+		canaryStatus:    nil,
+		expectedStatus:  createFailedStatus(PredictorHostnameUnknown, PredictorMissingMessage),
+		expectedService: nil,
+	}, {
+		name: "found predictor",
+		defaultStatus: &v1alpha2.ComponentStatusMap{
+			constants.Predictor: &v1alpha2.StatusConfigurationSpec{
+				Hostname: predictorHostname,
+			},
+		},
+		canaryStatus: nil,
+		expectedStatus: &v1alpha2.VirtualServiceStatus{
+			Status: duckv1beta1.Status{
+				Conditions: duckv1beta1.Conditions{{
+					Type:   v1alpha2.RoutesReady,
+					Status: corev1.ConditionTrue,
+				}},
+			},
+			URL:           expectedURL,
+			DefaultWeight: 100,
+		},
+		expectedService: &istiov1alpha3.VirtualService{
+			ObjectMeta: metav1.ObjectMeta{Name: "my-model"},
+			Spec: istiov1alpha3.VirtualServiceSpec{
+				Hosts:    []string{"my-model.myDomain"},
+				Gateways: []string{"someIngressGateway"},
+				HTTP: []istiov1alpha3.HTTPRoute{
+					{
+						Match: []istiov1alpha3.HTTPMatchRequest{{URI: &istiov1alpha1.StringMatch{Prefix: "/v1/models/my-model:predict"}}},
+						Route: []istiov1alpha3.HTTPRouteDestination{
+							{
+								Destination: istiov1alpha3.Destination{Host: "someIngressServiceName"},
+								Weight:      100,
+								Headers: &istiov1alpha3.Headers{
+									Request: &istiov1alpha3.HeaderOperations{Set: map[string]string{"Host": predictorHostname}},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}, {
+		name: "missing canary predictor",
+		defaultStatus: &v1alpha2.ComponentStatusMap{
+			constants.Predictor: &v1alpha2.StatusConfigurationSpec{
+				Hostname: predictorHostname,
+			},
+		},
+		canaryStatus: &v1alpha2.ComponentStatusMap{
+			constants.Predictor: nil,
+		},
+		expectedStatus:  createFailedStatus(PredictorStatusUnknown, PredictorMissingMessage),
+		expectedService: nil,
+	}, {
+		name: "canary predictor no hostname",
+		defaultStatus: &v1alpha2.ComponentStatusMap{
+			constants.Predictor: &v1alpha2.StatusConfigurationSpec{
+				Hostname: predictorHostname,
+			},
+		},
+		canaryStatus: &v1alpha2.ComponentStatusMap{
+			constants.Predictor: &v1alpha2.StatusConfigurationSpec{},
+		},
+		expectedStatus:  createFailedStatus(PredictorHostnameUnknown, PredictorMissingMessage),
+		expectedService: nil,
```

Comment:
I've seen a convention where you leave out nil values from the map and let it default.

Reply:
um - nice to have it explicit in this case?